pyccel/parser/syntax/openmp.py

```diff
@@ -31,142 +35,547 @@ def __init__(self, **kwargs):
+                             OmpTaskyieldConstruct, OmpTaskConstruct, OmpFlushConstruct,
+                             OmpCancelConstruct, OmpTargetConstruct, OmpTeamsConstruct,
+                             OmpDistributeConstruct, OmpSectionConstruct, OmpSectionsConstruct)):
             return stmt.expr
         else:
             raise TypeError('Wrong stmt for OpenmpStmt')
 
+def check_get_clauses(name, valid_clauses, clauses):
+    txt = ''
+    for clause in clauses:
+        if isinstance(clause, valid_clauses):
+            txt = '{0} {1}'.format(txt, clause.expr)
+        else:
+            raise TypeError('Wrong clause', type(clause).__name__,'for', type(name).__name__)
+    return txt
+
 class OmpParallelConstruct(BasicStmt):
-    """Class representing a ."""
+    """Class representing a parallel construct."""
     def __init__(self, **kwargs):
         """
         """
-        self.clauses = kwargs.pop('clauses')
-
-        super(OmpParallelConstruct, self).__init__(**kwargs)
+        self.clauses  = kwargs.pop('clauses')
+        self.combined = kwargs.pop('combined')
+
+        _valid_clauses = _valid_parallel_clauses
+
+        self.com = None
+        if isinstance(self.combined, OmpForSimd):
+            _valid_clauses = _valid_clauses + _valid_loop_clauses
+            if 'simd' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_simd_clauses
+            self.com = self.combined.expr
+        if isinstance(self.combined, OmpMaskedTaskloop):
+            _valid_clauses = _valid_clauses + (OmpFilter,)
+            if 'simd' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_simd_clauses
+            if 'taskloop' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_taskloop_clauses
+            self.com = self.combined.expr
+        if isinstance(self.combined, OmpPSections):
+            _valid_clauses = _valid_clauses + _valid_sections_clauses
+            self.com = self.combined.expr
+
+        self.txt = check_get_clauses(self, _valid_clauses, self.clauses)
+
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         if DEBUG:
             print("> OmpParallelConstruct: expr")
 
-        _valid_clauses = (OmpNumThread, \
-                         OmpDefault, \
-                         OmpPrivate, \
-                         OmpShared, \
-                         OmpFirstPrivate, \
-                         OmpCopyin, \
-                         OmpReduction, \
-                         OmpProcBind)
-
-        txt = 'parallel'
-        for clause in self.clauses:
-            if isinstance(clause, _valid_clauses):
-                txt = '{0} {1}'.format(txt, clause.expr)
-            else:
-                raise TypeError('Wrong clause for OmpParallelConstruct')
-
-        return OMP_Parallel_Construct(txt)
+        return OMP_Parallel_Construct(self.txt, self.com)
```

Comment:
Is there a reason why each call to `OmpParallelConstruct.expr` should create a new `OMP_Parallel_Construct` instance?